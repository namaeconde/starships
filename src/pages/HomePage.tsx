import React, { useEffect, useState } from "react";
import Button from "@mui/material/Button";
import { color } from "../theme";
import Box from "@mui/system/Box";
import Page from "../components/Page";
import { makeStyles } from "@mui/styles";
import Typography from "@mui/material/Typography";
import Grid from "@mui/material/Grid";
import StarshipCard from "../components/StarshipCard";
import Pagination from "../components/Pagination";
import Skeleton from '@mui/material/Skeleton';

const SWAPI_STARSHIPS_URL = "https://swapi.dev/api/starships";

function SideBar(): JSX.Element {
  const { sidebarButton } = makeStyles(() => ({
    sidebarButton: {
      fontSize: "24px",
      backgroundColor:`${color.racing_green}`,
      borderRadius:"20px",
      textAlign: "center"
    }
  }))();

  return (
    <>
      <Button className={ sidebarButton }>
          View Favorites
      </Button>
    </>
  )
}

function Body({ title, list }: any ): JSX.Element {
  return (
    <>
      <Box sx={{
        display: 'flex',
        justifyContent: 'flex-start'
      }}>
        <Typography fontWeight={700} fontSize={48}>{title}</Typography>
      </Box>
      {
        list ? <Box sx={{ width: '100%' }}>
          {
            list.length > 0 ?
                <Grid container rowSpacing={2} columnSpacing={2}>
                {
                  list.map((item: any, _index: number) => {
                    return (
                      <Grid key={_index} item xs={12} sm={6}>
                      <StarshipCard key={_index} starship={item}/>
                      </Grid>
                    )
                  })
                }
              </Grid> :
              <Typography>No starships found.</Typography>
          }
        </Box> :
        <Box sx={{ width: '100%' }}>
          {
            Array.from(new Array(10)).map((item: number, index: number) => (
              <Skeleton key={index}/>
            ))
          }
        </Box>
      }
    </>
  )
}

async function getStarshipsData(url: string) {
  return await fetch(url)
    .then(response => response.json())
    .then(data => {
      return data;
    })
}

export default function HomePage() {
  const [ previousUrl, setPreviousUrl ] = useState(null);
  const [ nextUrl, setNextUrl] = useState(SWAPI_STARSHIPS_URL);
  const [ starshipList, setStarshipList ] = useState(null);

  useEffect(() => {
    getStarshipsData(SWAPI_STARSHIPS_URL)
    .then(data => {
      const { results, next, previous } = data;
      setStarshipList(results);
      setNextUrl(next);
      setPreviousUrl(previous);
    })
  }, [])

  const onNextClick = nextUrl ? () => {
    setStarshipList(null);
    getStarshipsData(nextUrl)
      .then(data => {
        const { results, next, previous } = data;
        setStarshipList(results);
        setNextUrl(next);
        setPreviousUrl(previous);
      })
  } : null

  const onPreviousClick = previousUrl ? () => {
    setStarshipList(null);
    getStarshipsData(previousUrl)
      .then(data => {
        const { results, next, previous } = data;
        setStarshipList(results);
        setNextUrl(next);
        setPreviousUrl(previous);
      })   
  } : null

  return (
    <Page>
      {{
        sidebar: <SideBar />,
        body: <Body title="Starship List" list={starshipList}/>,
        footer: starshipList ? <Pagination onNextClick={ onNextClick } onPreviousClick={ onPreviousClick }/> : null
      }}
    </Page>
  )
}