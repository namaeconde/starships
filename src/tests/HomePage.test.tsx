import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import HomePage from '../pages/HomePage';

import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

const initialState = { favoriteList: { value: [] } }
const mockStore = configureStore()
let store

test('Header renders', () => {
  store = mockStore(initialState);
  const component = renderer.create(<Provider store={store}><Router><HomePage /></Router></Provider>,);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
