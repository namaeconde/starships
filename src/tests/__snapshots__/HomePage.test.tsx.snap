// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders 1`] = `
<div
  className="MuiBox-root css-8atqhb"
>
  <div
    className="MuiBox-root css-l10m55"
  >
    <div
      className="MuiBox-root css-175rvmj"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <div
          className="MuiAvatar-root MuiAvatar-square css-ivlicw-MuiAvatar-root"
        >
          <img
            alt="Starship"
            className="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="/assets/home_logo.png"
          />
        </div>
      </a>
    </div>
    <div
      className="makeStyles-container-1 MuiBox-root css-vq34d8"
    >
      <a
        className="button muted-button"
        href="/favorites"
        onClick={[Function]}
        style={
          Object {
            "textDecoration": "none",
          }
        }
      >
        <button
          className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root makeStyles-sidebarButton-3 css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          View favorites
        </button>
      </a>
    </div>
    <div
      className="makeStyles-main-2 MuiBox-root css-1xkyxdb"
    >
      <div
        className="MuiBox-root css-uj0oyz"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-15ixj63-MuiTypography-root"
        >
          Starship List
        </p>
      </div>
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined MuiBox-root css-1vtca5r-MuiGrid-root"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiBox-root css-es3gb9-MuiGrid-root"
        >
          <span
            className="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
            style={
              Object {
                "height": undefined,
                "width": undefined,
              }
            }
          />
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiBox-root css-es3gb9-MuiGrid-root"
        >
          <span
            className="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
            style={
              Object {
                "height": undefined,
                "width": undefined,
              }
            }
          />
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiBox-root css-es3gb9-MuiGrid-root"
        >
          <span
            className="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
            style={
              Object {
                "height": undefined,
                "width": undefined,
              }
            }
          />
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiBox-root css-es3gb9-MuiGrid-root"
        >
          <span
            className="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
            style={
              Object {
                "height": undefined,
                "width": undefined,
              }
            }
          />
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiBox-root css-es3gb9-MuiGrid-root"
        >
          <span
            className="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
            style={
              Object {
                "height": undefined,
                "width": undefined,
              }
            }
          />
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiBox-root css-es3gb9-MuiGrid-root"
        >
          <span
            className="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
            style={
              Object {
                "height": undefined,
                "width": undefined,
              }
            }
          />
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiBox-root css-es3gb9-MuiGrid-root"
        >
          <span
            className="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
            style={
              Object {
                "height": undefined,
                "width": undefined,
              }
            }
          />
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiBox-root css-es3gb9-MuiGrid-root"
        >
          <span
            className="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
            style={
              Object {
                "height": undefined,
                "width": undefined,
              }
            }
          />
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiBox-root css-es3gb9-MuiGrid-root"
        >
          <span
            className="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
            style={
              Object {
                "height": undefined,
                "width": undefined,
              }
            }
          />
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiBox-root css-es3gb9-MuiGrid-root"
        >
          <span
            className="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
            style={
              Object {
                "height": undefined,
                "width": undefined,
              }
            }
          />
        </div>
      </div>
    </div>
    <div
      className="makeStyles-container-1 MuiBox-root css-1icjmr2"
    />
  </div>
</div>
`;
